src/lib/http.coffee

```diff
@@ -40,42 +113,26 @@ module.exports = (dnschain) ->
                 @log.debug 'shutting down!'
                 @server.close cb
 
-        callback: (req, res, cb) ->
-            path = S(url.parse(req.url).pathname).chompLeft('/').s
-            options = url.parse(req.url, true).query
-            @log.debug gLineInfo('request'), {path:path, options:options, url:req.url}
-
-            notFound = =>
-                res.writeHead 404,  'Content-Type': 'text/plain'
-                res.write "Not Found: #{path}"
-                res.end()
-                cb()
-
-            [...,resolverName] =
-                if S(header = req.headers.blockchain || req.headers.host).endsWith('.dns')
-                    S(header).chompRight('.dns').s.split('.')
-                else
-                    ['none']
-
-            if not (resolver = @dnschain.chains[resolverName])
-                @log.warn gLineInfo('unknown blockchain'), {host: req.headers.host, blockchainHeader: req.headers.blockchain, remoteAddress: req.connection.remoteAddress}
-                res.writeHead 400, 'Content-Type': 'text/plain'
-                res.write "No Blockchain Found: #{resolverName}"
-                res.end()
-                cb()
-                return
-
-            if not resolver.validRequest path
-                @log.debug gLineInfo("invalid request: #{path}")
-                return notFound()
-
-            @dnschain.cache.resolveBlockchain resolver, path, options, (err,result) =>
+        callback: (req, res, args) ->
+            [datastoreName, resourceName, propOrAction] = args
+            if not (datastore = @dnschain.chains[datastoreName])
+                return @sendErr req, res, 400, "Unsupported datastore: #{datastoreName}"
+            if not (resourceFn = datastore.resources[resourceName])
+                return @sendErr req, res, 400,"Unsupported resource: #{resourceName}"
+            # TODO: deal with datastore.validRequest
```

Comment:
Please re-add this check. Notice that it got removed (see the red stuff at the left, [line 68](<#diff-5b98f1aa7adc82156f388fc5882a2a69L68>)).